### PR TITLE
Add implementation and tests for datetime.deltatime

### LINKED ISSUE #15: Pair2: Test and implement the chosen datetime work unit
For the work unit picked in #13,  "Add TranspilingTestsCases for your work unit and gradually add implementaions. Try to cover all conceivable inputs, edge cases and exceptions (if in doubt, refer to the Python documentation and follow their API specification). Finally, make yourself familiar with the [VOC guidelines](https://pybee.org/contributing/how/process/) regarding [code documentation](https://voc.readthedocs.io/en/latest/reference/signatures.html) and [code style](https://voc.readthedocs.io/en/latest/how-to/contribute.html#running-the-code-style-checks) and follow them closely in your implementations. "

# Chosen operations
| Operation                              | Result                                                                                |
|----------------------------------------|---------------------------------------------------------------------------------------|
| `t1  =  t2  +  t3`                     | Sum of t2 and t3. Afterwards t1-t2 == t3 and t1-t3 == t2 are true. (1)                |
| `t1  =  t2  *  i  or  t1  =  i  *  t2` | Delta multiplied by an integer. Afterwards t1 // i == t2 is true, provided i  !=  0 . |

# Checklist 

datetime.TimeDelta
- [x] constructor (args)
- [x] constructor (kwargs)
- class and instance attributes
    - [x] timedelta.days
    - [x] timedelta.seconds
    - [x] timedelta.microseconds
    - [x] timedelta.min
    - [x] timedelta.max
    - [x] timedelta.resoltion
- pick two operations 
   - [x] Addition
    - [x] Mulitplication
- [x] total_seconds()

Submit PR as with code review instructions! (i.e., how to run tests, what to look for, and so on)

### PR DESCRIPTION
**Pull request by rehnarama**
Fixes #15 

Squashed following commits
```
Add passing total_seconds test and implementation

Add day constructor without positional argument

Add microsecond support for days timedelta

Refactor

Move rounding of microseconds to before conversion to seconds

Add kwargs to timedelta function signature

Add passing seconds argument test

Add check for overflow error in days

Add tests for microsecond agrument

Add tests for rest of arguments

Add tests for instance attributes

To fix the total_seconds method we needed better precision, thus we use BigDecimal

Add test and implementation of kwargs

Add failing class attributes

Fix the failing class attributes by using instance of the class

Fix style errors from flake8

Make class attribute test look better

Multiplication and Addition for TimeDelta (#33)

* Implement timedelta multiplication with float and int

* Fix timedelta not rounding Half to even

* Implement addition for timedelta

Fix flake8 style checks
```
```
Co-authored-by: Adam Inersjö <adam.inersjo.2778@student.uu.se>
Co-authored-by: Joel Westerlund <jowe3751@student.uu.se>
Co-authored-by: Michael Rehn <rehn.michael@outlook.com>
```

Run test with 
`ant test && python setup.py test -s tests.stdlib.test_datetime.DatetimeModuleTests`

### CODE REVIEW

File: python/common/org/python/stdlib/datetime/timedelta.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
+package org.python.stdlib.datetime;
+
+import java.math.BigDecimal;
+
+public class timedelta extends org.python.types.Object {
+
```
Comment (JonasNorlinder) [inline python/common/org/python/stdlib/datetime/timedelta.java]: According to their convention there should not be any empty space here.

File: python/common/org/python/stdlib/datetime/timedelta.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    private static final long MAX_DAYS = 999999999;
+    private static final long MAX_SECONDS = 60 * 60 * 24 - 1;
+    private static final long MAX_MICROS = 1000000 - 1;
+
+    @org.python.Attribute
+    public final static org.python.Object min = new timedelta(-MAX_DAYS, 0, 0);
+    @org.python.Attribute
+    public final static org.python.Object max = new timedelta(MAX_DAYS, MAX_SECONDS, MAX_MICROS);
+    @org.python.Attribute
+    public final static org.python.Object resolution = new timedelta(0, 0, 1);
```
Comment (JonasNorlinder) [inline python/common/org/python/stdlib/datetime/timedelta.java]: No magic numbers.

```suggestion
    private static final long DEFAULT_DAYS = 0;
    private static final long DEFAULT_SECONDS= 0;
    private static final long DEFAULT_MICROS = 1;
    public final static org.python.Object resolution = new timedelta(DEFAULT_DAYS , DEFAULT_SECONDS, DEFAULT_MICROS);
```

File: python/common/org/python/stdlib/datetime/timedelta.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+    private timedelta(long days, long seconds, long microseconds) {
+        this((double) days, (double) seconds, (double) microseconds);
+    }
+
+    private timedelta(double days, double seconds, double microseconds) {
+        normalize(days, seconds, microseconds);
+    }
+
+    @org.python.Method(__doc__ = "Creates empty timedelta", default_args = { "iterable" }, kwargs = "kwargs")
```
Comment (JonasNorlinder) [inline python/common/org/python/stdlib/datetime/timedelta.java]: Why is `default_args = { "iterable" }`? `iterable` is never used

File: python/common/org/python/stdlib/datetime/timedelta.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+    private void normalize(double days, double seconds, double microseconds) {
+        // convert decimals to smaller units e.g. 1.5 days => 1 days and 43200 seconds
+        seconds += (days % 1) * 60 * 60 * 24;
+        days -= days % 1;
+
+        microseconds += (seconds % 1) * 1e6;
+        seconds -= seconds % 1;
+
+        // round the answer to whole microseconds
```
Comment (JonasNorlinder) [inline python/common/org/python/stdlib/datetime/timedelta.java]: Is this inline comment really necessary? I think the method name is self-explanatory.
```suggestion
```

File: python/common/org/python/stdlib/datetime/timedelta.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        // convert many seconds to days
+        days += (long) (seconds / (24 * 60 * 60));
+        seconds -= (long) (seconds / (24 * 60 * 60)) * 24 * 60 * 60;
+
+        if (microseconds < 0) {
+            days -= 1;
+            seconds += MAX_SECONDS;
+            microseconds += MAX_MICROS + 1;
+
+            // IF we overflow MAX_SECONDS again
```
Comment (JonasNorlinder) [inline python/common/org/python/stdlib/datetime/timedelta.java]: Typo.

```suggestion
            // if we overflow MAX_SECONDS again
```

File: python/common/org/python/stdlib/datetime/timedelta.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+                return (long) Math.ceil(d);
+            }
+        }
+        return Math.round(d);
+    }
+
+    @org.python.Method(__doc__ = "Return the total number of seconds contained in the duration.\n"
+            + " Equivalent to td / timedelta(seconds=1). For interval units \n"
+            + "other than seconds, use the division form directly \n"
+            + "(e.g. td / timedelta(microseconds=1)).", args = {})
```
Comment (JonasNorlinder) [inline python/common/org/python/stdlib/datetime/timedelta.java]: According to convention for `__doc__` should always end with '\n'
```suggestion
            + "(e.g. td / timedelta(microseconds=1)).\n", args = {})
```

File: python/common/org/python/stdlib/datetime/timedelta.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    public org.python.Object total_seconds() {
+        // Regular double precision isn't enough! We need the big stuff
```
Comment (JonasNorlinder) [inline python/common/org/python/stdlib/datetime/timedelta.java]: Maybe something more descriptive rather than repeat what we can see by reading the code.
```suggestion
        // Python handles doubles larger than the standard double in Java, thus we have to use BigDecimal
```

File: python/common/org/python/stdlib/datetime/timedelta.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        BigDecimal daysInSeconds = BigDecimal.valueOf(days.value).multiply(BigDecimal.valueOf(24 * 60 * 60));
+        BigDecimal bigSeconds = BigDecimal.valueOf(seconds.value);
+        BigDecimal microsecondsInSeconds = BigDecimal.valueOf(microseconds.value).divide(BigDecimal.valueOf(1e6));
+        BigDecimal total = daysInSeconds.add(bigSeconds).add(microsecondsInSeconds);
+        return new org.python.types.Float(total.doubleValue());
+    }
+
+    private long getIntValue(org.python.Object obj) {
```
Comment (JonasNorlinder) [inline python/common/org/python/stdlib/datetime/timedelta.java:135]: Just for reference, this is a bad pattern, but I get why you implemented it: `getIntValue()` should have been implemented in `org.python.types.Int`. The risk is now that each user of `org.python.types.Int` will implement its own getIntValue()

File: python/common/org/python/stdlib/datetime/timedelta.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        return ((org.python.types.Int) obj.__int__()).value;
+    }
+
+    private double getFloatvalue(org.python.Object obj) {
```
Comment (JonasNorlinder) [inline python/common/org/python/stdlib/datetime/timedelta.java:139]: Same goes here. This is not the responsibility of `timedelta`

File: python/common/python/datetime.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
+package python;
+
+@org.python.Module(
+    __doc__ = "This module provides various functions to datetime."
```
Comment (JonasNorlinder) [inline python/common/python/datetime.java]:
```suggestion
     __doc__ = "This module provides various functions to datetime.\n" 
```

**Review by JonasNorlinder — CHANGES_REQUESTED** [review]
Generally, good work! I have some minor suggestions.

File: python/common/org/python/stdlib/datetime/timedelta.java (hunk 8 above)
Comment (rehnarama) [reply python/common/org/python/stdlib/datetime/timedelta.java:135]: Very valid comment, this is very smelly code! Is it worth fixing now, or should we let it be?

File: python/common/org/python/stdlib/datetime/timedelta.java (hunk 9 above)
Comment (rehnarama) [reply python/common/org/python/stdlib/datetime/timedelta.java:139]: Very valid comment, this is very smelly code! Is it worth fixing now, or should we let it be?

File: python/common/org/python/stdlib/datetime/timedelta.java (hunk 8 above)
Comment (JonasNorlinder) [reply python/common/org/python/stdlib/datetime/timedelta.java:135]: If you have time it would be great, but it's up to you

**Review by JonasNorlinder — APPROVED** [review]
